chia/clvm/spend_sim.py

```diff
@@ -0,0 +1,299 @@
+import aiosqlite
+
+from typing import Optional, List, Dict, Tuple, Any
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program, SerializedProgram
+from chia.util.ints import uint64, uint32
+from chia.util.hash import std_hash
+from chia.util.errors import Err
+from chia.util.db_wrapper import DBWrapper
+from chia.types.coin_record import CoinRecord
+from chia.types.spend_bundle import SpendBundle
+from chia.types.generator_types import BlockGenerator
+from chia.types.mempool_inclusion_status import MempoolInclusionStatus
+from chia.types.coin_spend import CoinSpend
+from chia.full_node.bundle_tools import simple_solution_generator
+from chia.full_node.mempool_manager import MempoolManager
+from chia.full_node.coin_store import CoinStore
+from chia.full_node.mempool_check_conditions import get_puzzle_and_solution_for_coin
+from chia.consensus.default_constants import DEFAULT_CONSTANTS
+from chia.consensus.coinbase import create_pool_coin, create_farmer_coin
+from chia.consensus.block_rewards import calculate_pool_reward, calculate_base_farmer_reward
+from chia.consensus.cost_calculator import NPCResult
+
+"""
+The purpose of this file is to provide a lightweight simulator for the testing of Chialisp smart contracts.
+
+The Node object uses actual MempoolManager, Mempool and CoinStore objects, while substituting FullBlock and
+BlockRecord objects for trimmed down versions.
+
+There is also a provided NodeClient object which implements many of the methods from chia.rpc.full_node_rpc_client
+and is designed so that you could test with it and then swap in a real rpc client that uses the same code you tested.
+"""
+
+
+class TestFullBlock:
+    def __init__(self, generator: Optional[BlockGenerator], height: uint32):
+        self.height = height  # Note that height is not on a regular FullBlock
+        self.transactions_generator = generator
+
+
+class TestBlockRecord:
+    def __init__(self, rci: List[Coin], height: uint32, timestamp: uint64):
+        self.reward_claims_incorporated = rci
+        self.height = height
+        self.prev_transaction_block_height = uint32(height - 1) if height > 0 else 0
+        self.timestamp = timestamp
+        self.is_transaction_block = True
+        self.header_hash = std_hash(bytes(height))
+
+
+class SpendSim:
+
+    connection: aiosqlite.Connection
+    mempool_manager: MempoolManager
+    block_records: List[TestBlockRecord]
+    blocks: List[TestFullBlock]
+    timestamp: uint64
+    block_height: uint32
+
+    @classmethod
+    async def create(cls):
```

Comment:
Would be slightly nicer to bring DEFAULT_CONSTANTS in through an argument to `create`

```suggestion
    async def create(cls, sim_constants: ConsensusConstants):
```